src/nbl/asset/interchange/CGLILoader.h

```diff
@@ -20,11 +20,12 @@ namespace asset
 //! Texture loader capable of loading in .ktx, .dds and .kmg file extensions
 class CGLILoader final : public asset::IAssetLoader
 {
+	core::smart_refctd_ptr<system::ISystem> m_system;
```

Comment:
you could move it to IAssetLoader, all loaders will use ISystem anyway

Reply:
@Crisspl I already brought it up, but I'm of the opinion that the file should keep a smart pointer to the `system` that created it (and it needs for performing async reads and writes)

Because if someone gets the idea to create multiple system objects and multiple files, this turns into a clusterfuck (wrong system reading a file that doesn't belong to it)

Also file should keep the system alive while the file is alive, so a backward pointer makes sense.

Reply:
yeah, but in ISystem.h there needs some declaration like `smart_refctd_ptr<IFile>` and in IFile then would need to be a declaration like `smart_refctd_ptr<ISystem>`, but you cannot do this with forward-declaring the classes since at that point it must be known that T in smart_refctd_ptr<T> is derivative of IReferenceCounted. So it's basically impossible to do easy way (well, unless we get rid of static_assert in smart_refctd_ptr). So we postponed it to not lose time on not that important things

Reply:
> So it's basically impossible to do easy way (well, unless we get rid of static_assert in smart_refctd_ptr).

Is it just the assert that's causing this? not the function definitions, etc?

Yeah this is kinda annoying, can we move the assert somewhere else?

Reply:
Hm, right, whole class definition must be known to use it with smart_refctd_ptr, with or without static_assert. So eh, it's basically impossible to do

Reply:
ok so the solution here is to split `ISystem` and to inherit from `IFileOperator` and `IFileFactory` where:
- File Factory would hand out the `smart_refctd_ptr<IFile>` and the files need not know about the factory
- File Operator's protected methods would operate on "dumb pointers" to `IFile`s and its the only reason why an `IFile` wants to hold a smartpointer to an `IFileOperator` (to ensure operator is still alive when it wants to use the operator to implement its own methods)

Reply:
do you think both of them would have a thread? Or they share one thread under the hood?
Anyway.. IMO it's easier and cleaner to just have IFile hold ISystem with raw pointer (include ISystem in .cpp file and use grab/drop directly). Maybe even have a base class for backward linking (an object holding its creator) like
```
// IBackwardLink.h
// have it templated for case when theres more than 1 backward links needed (possible?)
template <typename T> // T is some forward-declared class
class IBackwardLink
{
explicit IBackwardLink(T* c);
~IBackwardLink();
T* getCreator() { return creator; }
private:
T* creator;
};

//IBackwardLink.cpp
// have to explicitely instantiate (force compiling them) those ctor and dtor when needed
template <typename T>
IBackwardLink<T>::IBackwardLink(T* c) : creator(c) { c->grab(); }
template<typename T>
IBackwardLink<T>::~IBackwardLink() { creator->drop(); }
```

A little weird but DRY and dont have to split ISystem which might be problematic. IFileFactory would have to have pointer to IFileOperator to give it to files being created (if you want files to be able to be read and written to w/o direct usage of IFileOperator API as you said). AND often it's needed to be able to both create/open and operate on files, so they would passed in pair to things (IAssetManager, IGLSLCompiler, more?)
I'm team IBackwardLink

Reply:
> do you think both of them would have a thread? Or they share one thread under the hood?
Well the most derived one (operator) would hold most of the stuff, the base (factory) is just an interface.

> Anyway.. IMO it's easier and cleaner to just have IFile hold ISystem with raw pointer (include ISystem in .cpp file and use grab/drop directly).
Would the usage of the backward link util class be via inheritance, or composition?
I'm liking this direction, i'd call it `IFactoryKeepalive` or `IFactoryLifetimeSupport` though.

In theory definitely better than forced splits because of how `smart_refctd_ptr` works.

But there's one huge caveat, this is essentially a `smart_refctd_ptr` but the constructor and destructor are not inlined, but defined in a different TU and explicitly instantiated for every class you might want to use them with.